### ISSUE TITLE
Array of type "date" not working

### ISSUE THREAD

**Issue by bvenky27** [strongloop/loopback-datasource-juggler#1632]
When I declare a property of type Array of Dates (`"type" : ["date"]`) then while inserting the Array of Dates in the database, it is storing as Array of Strings with present time but not as Array of Date objects.

I found a bug in  `\lib\list.js` file in `toItem()` where there is some error in logic.

```js
  List.prototype.toItem = function(item) {
    return isClass(this.itemType) ? new this.itemType(item) : this.itemType(item);
  };
```

I changed the above function to something like below, just swaped if and else part and it is working fine:

```js
  List.prototype.toItem = function(item) {
    return isClass(this.itemType) ? this.itemType(item) : new this.itemType(item);
  };
```

Should I do a pull request or will you fix it? please let me know

**Comment by bajtos** [comment]
Thank you for reporting the issue. IIUC, the problem you are trying to solve is that `toItem` does not convert a date string (e.g. `"2018-01-01T00:00:00.000Z"`) to a Date instance, but returns a string instead.

```
$ node
> new Date('2018-01-01T00:00:00.000Z')
2018-01-01T00:00:00.000Z
> Date('2018-01-01T00:00:00.000Z')
'Tue Oct 09 2018 12:34:11 GMT+0200 (Central European Summer Time)'
```

I am afraid your proposed change is not fixing the root cause of the problem and will break code that's using an ES6 class for `itemType`.

```
$ node
> class Foo {}
undefined
> isClass(Foo)
true
> Foo('value')
TypeError: Class constructor Foo cannot be invoked without 'new'
```

What we need to do instead: fix `isClass` function to correctly recognize built-in classes like `Date` and `RegExp`.

We don't have bandwidth to make this change ourselves, but I am happy to help you along the way if you decide to contribute this fix. Just don't forget to add new tests to verify your implementation and prevent regressions in the future, see [test/list.test.js](https://github.com/strongloop/loopback-datasource-juggler/blob/c95a969af7382d3c03b60c4df42d09c8785617ea/test/list.test.js).

Please read our [Contributing code](https://loopback.io/doc/en/contrib/code-contrib.html) guide to better understand the coding style, commit message format and other conventions we are following.

**Comment by bvenky27** [comment]
@bajtos yes, my proposed change breaks the code. I found out that the problem is not with the isClass function. It recognizes the built-in classes like Date correctly.

The problem is with the 'this.itemType(item)'. Here when the 'itemType' is of type 'Number' then Number(123) gives result as 123, a number and its the same for String(<value>) too, it will return string. But its not the case for 'Date(<stringValue>)'. This will return a string with Present date information but not the Date instance of specified <stringValue>.
For example: Date('1995-12-17T03:24:00') give " Fri Jan 25 2019 10:51:56 GMT-0500 (Eastern Standard Time" which is not the expected value.
but new Date('1995-12-17T03:24:00') gives a Date object with 'Sun Dec 17 1995 03:24:00 GMT-0500 (Eastern Standard Time)'.

How about the change below? adding a special case for Date Object as it is failing only in that case?

List.prototype.toItem = function(item) {
    return (isClass(this.itemType) || (this.itemType.toString() == Date.toString())) ? new this.itemType(item) : this.itemType(item);
  };

**Comment by bajtos** [comment]
> How about the change below? adding a special case for Date Object as it is failing only in that case?
>
> ```js
> List.prototype.toItem = function(item) {
>   return (isClass(this.itemType) || (this.itemType.toString() == Date.toString())) ? new this.itemType(item) : this.itemType(item);
> };
> ```

This looks reasonable to me. Let's continue the discussion in https://github.com/strongloop/loopback-datasource-juggler/pull/1687.

**Comment by bajtos** [comment]
Fixed by #1702

**Comment by dhmlau** [comment]
Let me close it as resolved. Thanks @bajtos for pointing it out.
